### PR TITLE
Fix #2325: Build for 2.12.0-M4.

### LINKED ISSUE #2325: Upgrade to Scala 2.12.0-M4
Should replace 2.12.0-M3.

### PR DESCRIPTION
**Pull request by nicolasstucki**
Review by @sjrd

### CODE REVIEW

File: partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+# Bug #2309
+run/sammy_after_implicit_view.scala
+run/sammy_restrictions_LMF.scala
+run/sammy_vararg_cbn.scala
+run/t5568.scala
+run/t6318_primitives.scala
+
+# fails assert(x == y, x + " == " + y)
+# where (x: java.lang.Long = 5) and (y: scala.math.BigDecimal = 5)
+run/equality.scala
```
Comment (nicolasstucki) [inline partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt]: This one will need some investigation, it was in `WhitelistedTests.txt` for `2.12.0-M3`.

File: partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BlacklistedTests.txt [hunk 2]
```diff
@@ -598,6 +592,10 @@ run/t7521b.scala
 run/t8575b.scala
 run/t8575c.scala
 run/t8944c.scala
+run/t9535.scala
+run/t9437a
+run/t9437b
+
```
Comment (sjrd) [inline partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BlacklistedTests.txt]: Spurious second blank line.

File: partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# The tests in this file should pass but have never passed so far
+# use scala.tools.partest.scalajs.testunknownonly to only run tests
+# which are neither in BuglistedTests.txt, WhitelistedTests.txt or
+# BlacklistedTests.txt
+
+# Bug #2309
+run/sammy_after_implicit_view.scala
+run/sammy_restrictions_LMF.scala
+run/sammy_vararg_cbn.scala
+run/t5568.scala
```
Comment (sjrd) [inline partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt]: https://github.com/scala/scala/blob/2.12.x/test/files/run/t5568.scala
Uh? How is this related to #2309?

File: partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+run/t6318_primitives.scala
```
Comment (sjrd) [inline partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt]: https://github.com/scala/scala/blob/2.12.x/test/files/run/t6318_primitives.scala
Same here

**Comment by sjrd** [comment]
Could you separate the upgrade to scala-partest 1.0.13 in a separate (first) commit?

File: project/Build.scala [hunk 5]
```diff
@@ -1240,6 +1240,14 @@ object Build {
             fullName
         })),
 
+        /* 2.12.0-M4 introduced this bug, this should be removed in 2.12.0-M5
+         * with all references to "scalac.hasBoxedUnitBug" in the tests
+         */
+        javaOptions in Test ++= {
+          if (scalaVersion.value != "2.12.0-M4") Nil
+          else Seq("-Dscalac.hasBoxedUnitBug=ture")
```
Comment (sjrd) [inline project/Build.scala]: typo: `true`

**Comment by sjrd** [comment]
That's all.

File: partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BlacklistedTests.txt (hunk 2 above)
Comment (nicolasstucki) [inline partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BlacklistedTests.txt]: Removed

File: ci/checksizes.sh [hunk 6]
```diff
@@ -45,10 +45,10 @@ case $FULLVER in
     REVERSI_PREOPT_GZ_EXPECTEDSIZE=71000
     REVERSI_OPT_GZ_EXPECTEDSIZE=32000
     ;;
-  2.12.0-M3)
-    REVERSI_PREOPT_EXPECTEDSIZE=530000
-    REVERSI_OPT_EXPECTEDSIZE=123000
-    REVERSI_PREOPT_GZ_EXPECTEDSIZE=72000
+  2.12.0-M4)
+    REVERSI_PREOPT_EXPECTEDSIZE=593000
```
Comment (nicolasstucki) [inline ci/checksizes.sh]: The `REVERSI_PREOPT_EXPECTEDSIZE` increased by 12%. Is there any known reason for this?

Comment (sjrd) [inline ci/checksizes.sh]: Probably the fact that the optimizer is not yet able to multi-inline default method bridges.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/scalalib/ArrayBuilderTest.scala [hunk 7]
```diff
@@ -190,6 +191,8 @@ class ArrayBuilderTest {
   }
 
   @Test def Unit_zeros_inline(): Unit = {
+    assumeFalse("Can't run on 2.12.0-M4.",
+        System.getProperty("scalac.hasBoxedUnitBug") != null || executingInJVM)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/scalalib/ArrayBuilderTest.scala]: It does not make sense to prevent `executingInJVM` here, and later do `if (!executingInJVM)`.

From the changes about `ArrayBuilder.ofUnit` in Scala 2.12.0-M4, I would suggest to remove the `else` part of the `if (!executingInJVM)` instead of using `|| executingInJVM` in the assumption.

Comment (nicolasstucki) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/scalalib/ArrayBuilderTest.scala]: Done.

**Comment by sjrd** [comment]
Check partest coverage failed. Otherwise LGTM

File: partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt (hunk 1 above)
Comment (sjrd) [inline partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt]: Actually, `run/equality.scala` is broken on master. It started failing in the nightly of April 12.

Comment (sjrd) [inline partest-suite/src/test/resources/scala/tools/partest/scalajs/2.12.0-M4/BuglistedTests.txt]: #2335

**Comment by sjrd** [comment]
Nightly: https://scala-webapps.epfl.ch/jenkins/job/scalajs-matrix-build/3891/
(will fail because of #2335, but the 2.12.0-M4 jobs should pass.

**Comment by sjrd** [comment]
LGTM
All 2.12.0-M4 jobs passed.
